functorch/functorch/_src/partitioners.py

```diff
@@ -164,6 +174,100 @@ def default_partition(
 
     return _extract_fwd_bwd_modules(joint_module, saved_values)
 
+# This function expects a post-functionalizat FX GraphModule,
+# which means that all ops in the graph are functional, except for
+# some copy_() or as_strided_() nodes, that can show up at the end
+# of a graph who's module involves input mutations.
+#
+# This function extracts those input mutations into an opaque
+# submodule, called "input_mutation_submodule".
+#
+# We also update fx_g inplace to add the mutated inputs directly
+# as outputs in the graph.
+# Finally, this function returns a boolean list
+# telling AOTAutograd which of the inputs were actually mutated,
+# which correspond to actual inputs in the "input_mutation_submodule".
+#
+# AOTAutograd is responsible for splitting off the mutated inputs
+# after calling the compiled forward graph,
+# and running "input_mutation_submodule".
+def move_input_mutations_into_submodule(fx_g: fx.GraphModule) -> List[bool]:
+    mutable_nodes: List[Node] = []  # the copy_() and as_strided_() nodes, that we'll delete move to a submodule.
+    mutated_inputs: List[Node] = []
+    node_map: Dict[Node, Node] = {}
+    node_to_placeholder: Dict[Node, Node] = OrderedDict()  # mapping of nodes from old graph to placeholder in new submodule.
+    mutation_subgraph = fx.Graph()
+    model_input_to_mutation: Dict[Node: bool] = OrderedDict()
+
+    # First, default all model inputs to not having mutations
+    for n in fx_g.graph.nodes:
+        if n.op == 'placeholder':
+            model_input_to_mutation[n] = False
+
+    def remap_input(x):
+        if x in node_map:
+            return node_map[x]
+        if x not in node_to_placeholder:
+            # Add the copy_() mutated input argument as a submodule input.
+            placeholder_node = mutation_subgraph.placeholder(x.name, type_expr=x.type)
+            placeholder_node.meta = copy.copy(x.meta)
+            node_to_placeholder[x] = placeholder_node
+        return node_to_placeholder[x]
+
+    inpt_nodes = [n for n in fx_g.graph.nodes if n.op == 'placeholder']
+    for n in fx_g.graph.nodes:
+        if not isinstance(n.target, torch._ops.OpOverload):
+            continue
```

Comment:
@SherlockNoMad is the API ready for checking preconditions yet? The precondition here is that the graph is functionalized and it's all torch.ops calls

Reply:
Something like this? 

https://github.com/pytorch/pytorch/pull/83134